glutin/src/api/egl/sync.rs

```diff
@@ -0,0 +1,166 @@
+//! EGL Sync Fences.
+
+use std::ffi::c_void;
+use std::mem::MaybeUninit;
+use std::sync::Arc;
+use std::time::Duration;
+
+use glutin_egl_sys::egl::types::{EGLenum, EGLint};
+
+use super::display::DisplayInner;
+use super::{egl, ErrorKind, VERSION_1_5};
+use crate::error::Result;
+
+/// EGL sync object.
+#[derive(Debug, Clone)]
+pub struct Sync(pub(super) Arc<Inner>);
+
+impl Sync {
+    /// Insert this sync into the currently bound context.
+    ///
+    /// If the EGL version is not at least 1.5 or the `EGL_KHR_wait_sync`
+    /// extension is not available, this returns [`ErrorKind::NotSupported`].
+    ///
+    /// This will return [`ErrorKind::BadParameter`] if there is no currently
+    /// bound context.
+    pub fn wait(&self) -> Result<()> {
+        if self.0.display.version < VERSION_1_5
+            && !self.0.display.display_extensions.contains("EGL_KHR_wait_sync")
+        {
+            return Err(ErrorKind::NotSupported(
+                "Sync::wait is not supported if EGL_KHR_wait_sync isn't available",
+            )
+            .into());
+        }
+
+        if unsafe { self.0.display.egl.WaitSyncKHR(*self.0.display.raw, self.0.inner, 0) }
+            == egl::FALSE as EGLint
+        {
+            return Err(super::check_error().err().unwrap());
+        }
+
+        Ok(())
+    }
+
+    /// Query if the sync is already
+    pub fn is_signalled(&self) -> Result<bool> {
+        let status = unsafe { self.get_attrib(egl::SYNC_STATUS) }? as EGLenum;
+        Ok(status == egl::SIGNALED)
+    }
+
+    /// Block and wait for the sync object to be signalled.
+    ///
+    /// A timeout of [`None`] will wait forever. If the timeout is [`Some`], the
+    /// maximum timeout is [`u64::MAX`] - 1 nanoseconds. Anything larger will be
+    /// truncated. If the timeout is reached this function returns [`false`].
+    ///
+    /// If `flush` is [`true`], the currently bound context is flushed.
+    pub fn client_wait(&self, timeout: Option<Duration>, flush: bool) -> Result<bool> {
+        let flags = if flush { egl::SYNC_FLUSH_COMMANDS_BIT } else { 0 };
+        let timeout = timeout
+            .as_ref()
+            .map(Duration::as_nanos)
+            .map(|nanos| nanos.max(u128::from(u64::MAX)) as u64)
+            .unwrap_or(egl::FOREVER);
+
+        let result = unsafe {
+            self.0.display.egl.ClientWaitSyncKHR(
+                *self.0.display.raw,
+                self.0.inner,
+                flags as EGLint,
+                timeout,
+            )
+        } as EGLenum;
+
+        match result {
+            egl::FALSE => Err(super::check_error().err().unwrap()),
+            egl::TIMEOUT_EXPIRED => Ok(false),
+            egl::CONDITION_SATISFIED => Ok(true),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Export the fence's underlying sync fd.
+    ///
+    /// Returns [`ErrorKind::NotSupported`] if the sync is not a native fence.
+    ///
+    /// # Availability
```

Comment:
```suggestion
    /// ## Availability
```

I think glutin uses that amount of `##`?